crate/server/src/database/redis/permissions.rs

```diff
@@ -1,196 +1,131 @@
-                .query_async(&mut self.manager.clone())
-                .await
-                .map_err(FindexServerError::from)?
-        };
-
-        let serialized_value = &values.pop().ok_or_else(|| {
-            FindexServerError::Unauthorized(format!("No permission found for {user_id}"))
-        })?;
-
-        if serialized_value.is_empty() {
-            Ok(Permissions::default())
-        } else {
-            Permissions::deserialize(serialized_value).map_err(FindexServerError::from)
-        }
-    }
-
-    async fn get_permission(&self, user_id: &str, index_id: &Uuid) -> FResult<Permission> {
-        let _guard = self.lock.lock().await;
-        let permissions = self.get_permissions(user_id).await?;
-        let permission = permissions.get_permission(index_id).ok_or_else(|| {
-            FindexServerError::Unauthorized(format!(
-                "No permission for {user_id} on index {index_id}"
-            ))
-        })?;
-
-        Ok(permission.clone())
-    }
-
+    /// Grants a permission to a user for a specific index.
     async fn grant_permission(
         &self,
         user_id: &str,
         permission: Permission,
         index_id: &Uuid,
     ) -> FResult<()> {
-        let redis_key = user_id.as_bytes().to_vec();
+        let user_key = format!("{PERMISSIONS_PREFIX}:{user_id}");
+        let perm_number = u8::from(permission.clone());
 
-        let _guard = self.lock.lock().await;
-        let _p = self.get_permissions(user_id).await;
-
-        let permissions = match _p {
-            Ok(mut permissions) => {
-                debug!("permissions that are going to be set: {permissions:?}",);
-                permissions.grant_permission(*index_id, permission); // adds the "new" permission to the existing ones
-                permissions
-            }
-            Err(_) => Permissions::new(*index_id, permission),
-        };
-
-        let mut pipe = pipe();
-        let () = pipe
-            .atomic()
-            .set(&redis_key, permissions.serialize()?.as_slice())
-            .query_async(&mut self.manager.clone())
+        let _: () = self
+            .manager
+            .clone()
+            .hset(&user_key, index_id.to_string(), perm_number)
             .await
             .map_err(FindexServerError::from)?;
 
-        let mut pipe = redis::pipe();
-        let mut values: Vec<Vec<u8>> = pipe
-            .atomic()
-            .get(redis_key)
-            .query_async(&mut self.manager.clone())
+        trace!("Granted {permission:?} permission to {user_id} for index {index_id}");
+        Ok(())
+    }
+
+    #[instrument(ret(Display), err, skip(self), level = "trace")]
+    async fn get_permissions(&self, user_id: &str) -> FResult<Permissions> {
+        let user_redis_key = format!("{PERMISSIONS_PREFIX}:{user_id}");
+
+        let permissions: Permissions = self
+            .manager
+            .clone()
+            .hgetall::<_, HashMap<String, String>>(user_redis_key)
```

Comment:
You don't need a hash-map here and a vector will be faster.
```suggestion
            .hgetall::<_, Vec<(String, String)>>(user_redis_key)
```